src/mono/mono/mini/mini-llvm.c

```diff
@@ -7820,6 +7965,222 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 			break;
 		}
 
+		case OP_SSE2_PSLLDQ:
+		case OP_SSE2_PSRLDQ: {
+			LLVMBasicBlockRef bbs [16 + 1];
+			LLVMValueRef switch_ins;
+			LLVMValueRef value = lhs;
+			LLVMValueRef index = rhs;
+			LLVMValueRef phi_values [16 + 1];
+			LLVMTypeRef t = sse_i1_t;
+			int nelems = 16;
+			int i;
+			gboolean shift_right = (ins->opcode == OP_SSE2_PSRLDQ);
+
+			value = convert (ctx, value, t);
+
+			// No corresponding LLVM intrinsics
+			// FIXME: Optimize const count
+			for (i = 0; i < nelems; ++i)
+				bbs [i] = gen_bb (ctx, "PSLLDQ_CASE_BB");
+			bbs [nelems] = gen_bb (ctx, "PSLLDQ_DEF_BB");
+			cbb = gen_bb (ctx, "PSLLDQ_COND_BB");
+
+			switch_ins = LLVMBuildSwitch (builder, index, bbs [nelems], 0);
+			for (i = 0; i < nelems; ++i) {
+				LLVMAddCase (switch_ins, LLVMConstInt (LLVMInt32Type (), i, FALSE), bbs [i]);
+				LLVMPositionBuilderAtEnd (builder, bbs [i]);
+
+				int mask_values [16];
+				// Implement shift using a shuffle
+				if (shift_right) {
+					for (int j = 0; j < nelems - i; ++j)
+						mask_values [j] = i + j;
+					for (int j = nelems -i ; j < nelems; ++j)
+						mask_values [j] = nelems;
+				} else {
+					for (int j = 0; j < i; ++j)
+						mask_values [j] = nelems;
+					for (int j = 0; j < nelems - i; ++j)
+						mask_values [j + i] = j;
+				}
+				phi_values [i] = LLVMBuildShuffleVector (builder, value, LLVMGetUndef (t), create_const_vector_i32 (mask_values, nelems), "");
+				LLVMBuildBr (builder, cbb);
+			}
+			/* Default case */
+			LLVMPositionBuilderAtEnd (builder, bbs [nelems]);
+			phi_values [nelems] = LLVMConstNull (t);
+			LLVMBuildBr (builder, cbb);
+
+			LLVMPositionBuilderAtEnd (builder, cbb);
+			values [ins->dreg] = LLVMBuildPhi (builder, LLVMTypeOf (phi_values [0]), "");
+			LLVMAddIncoming (values [ins->dreg], phi_values, bbs, nelems + 1);
+			values [ins->dreg] = convert (ctx, values [ins->dreg], type_to_sse_type (ins->inst_c1));
+
+			ctx->bblocks [bb->block_num].end_bblock = cbb;
+			break;
+		}
+
+		case OP_SSE2_PSRAW_IMM:
+		case OP_SSE2_PSRAD_IMM:
+		case OP_SSE2_PSRLW_IMM:
+		case OP_SSE2_PSRLD_IMM:
+		case OP_SSE2_PSRLQ_IMM: {
+			LLVMBasicBlockRef bbs [64 + 1];
+			LLVMValueRef switch_ins;
+			LLVMValueRef value = lhs;
+			LLVMValueRef index = rhs;
+			LLVMValueRef phi_values [64 + 1];
+			IntrinsicId id;
+			int ncases;
+
+			// FIXME: Optimize const index case
+
+			switch (ins->opcode) {
+			case OP_SSE2_PSRAW_IMM:
```

Comment:
it stands for `ShiftRightArithmetic` which is `_mm_srai_epi16` which works with non-constant arguments normally https://godbolt.org/z/xWq6za why do we need to emit a huge switch for it?
Also I'd just throw an exception for non-cns arguments where intrinsics expect cns.

Reply:
E.g. https://github.com/dotnet/runtime/blob/master/src/mono/mono/mini/simd-intrinsics-netcore.c#L881-L885

Reply:
Here is what `ShiftRightArithmetic` emits now: https://gist.github.com/EgorBo/ea3f749413d573f5f1d4a522ff7cf841

Reply:
Looks like coreclr will just emit the non-immediate form of psraw when the shift amount can't be folded into a constant: https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4CS5AThaRBQMYT4AaAHAGwBudFlwES5KrQbBmrDhGFoRAZlwJsAYWwBvdNgO41ANWrtgAe0ZIEvADwyk/AHzYAsgApT5qzfuOXbAAbAAsIABpsACMAT2BqbDAoYABKXX1DTLgAdmwAZQhqBHw8kIoAM2BCCgBzEOAAQWZgEJpKdg9QiMTklKVMgF90AaA==

I agree that the switch doesn't seem necessary; we could just rely on LLVM to optimize the IR and select either the immediate or non-immediate form of these instructions as appropriate. https://godbolt.org/z/Zxagk4